### DISCUSSION TITLE
Custom wire:confirm directive doesn't work with async code

### DISCUSSION THREAD

**Discussion by gjm** [livewire/livewire#8413] [1. Help]
Hi.

I've made a custom `wire:confirm` directive to use Bootstrap modals to ask for confirmation. Since Bootstrap's modals are async I've made a wrapper function `bs_confirm` that returns a Promise and I can use `await` to make it pseudo-synchronous.

If I use like 

```html
<button wire:click="doSomething" wire:bs-confirm="Are you sure?">
    Do something
</button>
```

It works as expected except for the fact that, as soon as the modal opens, the request is made. Even thought I use `await` to make sure the promise is fulfilled.

I'm sure I'm missing something in the async code management but can't quite figure out what the problem is. If I place `console.log` before and after calling the `async` function with `await` it works as expected so I'm guessing it must be something to do with Livewire's handling of the events. 

This is my directive:

```js
Livewire.directive('bs-confirm', ({ el, directive, component, cleanup }) => {
    let content = directive.expression

    let onClick = async (e) => {
        if (! await bs_confirm(content)) {
            e.preventDefault()
            e.stopImmediatePropagation()
        }
    }

    el.addEventListener('click', onClick, { capture: true })

    cleanup(() => {
        el.removeEventListener('click', onClick)
    })
});
```

Can someone give me a help on this?

Thanks in advance.

**Comment by joshhanley** [comment]
@gjm `wire:click` will be firing. I would have a look at how Livewire's `wire:confirm` works and see how the `wire:click` is stopped/ ignored (it's all internal - a javascript click listener won't work directly).

https://github.com/livewire/livewire/blob/main/js/directives/wire-confirm.js

https://github.com/livewire/livewire/blob/dd253672b6f11b137bfc2d734b7b64b258483640/js/directives/wire-wildcard.js#L26-L35

**Reply by gjm** [reply]
I followed the documentation on directives (https://livewire.laravel.com/docs/javascript#registering-custom-directives) that shows how the `wire:confirm` directive is implemented. 

AFAIK the difference is that `confirm` actually blocks execution and the `await ...` doesn't even though it seems to if I do it elsewhere.

**Comment by CristhoferMF** [comment]
@gjm  Try this.
```js
Livewire.directive('bs-confirm', ({ el, directive, component, cleanup }) => {
    let content = directive.expression

    let onClick = async (e) => {
        e.preventDefault()
        e.stopImmediatePropagation()
        if (await bs_confirm(content)) {
          // is confirmed
         // remove this event
          el.removeEventListener('click', onClick, {
              capture: true
          })
        // call default event like: wire:click, submit ....
          el.click();
       // add this event
          el.addEventListener('click', onClick, {
              capture: true
          })
        }
    }

    el.addEventListener('click', onClick, { capture: true })

    cleanup(() => {
        el.removeEventListener('click', onClick, { capture: true })
    })
});
```
